hetmech/matrix.py

```diff
@@ -123,3 +124,55 @@ def copy_array(matrix, copy=True):
         mat_type = numpy.array
     matrix = mat_type(matrix, dtype=numpy.float64, copy=copy)
     return matrix
+
+
+def categorize(metapath):
+    """
+    Returns the classification of a given metapath as one of
+    a set of metapath types which we approach differently.
+
+    Parameters
+    ----------
+    metapath : hetio.hetnet.MetaPath
+
+    Returns
+    -------
+    classification : string
+        One of ['no_repeats', 'disjoint', 'BAAB', 'BABA', 'other']
+    """
+    metanodes = list(metapath.get_nodes())
+    repeated_nodes = {v for i, v in enumerate(metanodes) if
+                      v in metanodes[i + 1:]}
+
+    if not repeated_nodes:
+        return 'no_repeats'
+
+    repeats_only = [node for node in metanodes if node in repeated_nodes]
+
+    # Group neighbors if they are the same
+    grouped = [list(v) for k, v in itertools.groupby(repeats_only)]
+
+    # Handle multiple disjoint repeats, any number, ie. AA,BB,CC,DD,...
+    if len(grouped) == len(repeated_nodes):
+        return 'disjoint'
+
+    # Group [A, BB, A] or [A, B, A, B] into one
+    if len(repeats_only) - len(grouped) <= 1:
+        grouped = [[item for sublist in grouped for item in sublist]]
```

Comment:
is this the same as `[list(itertools.chain(*grouped))]`?

Is this also the same as `[repeats_only]`?